pkg/provisioner/ironic/ironic.go

```diff
@@ -1124,6 +1124,45 @@ func (p *ironicProvisioner) GetFirmwareSettings(includeSchema bool) (settings me
 	return settings, schema, nil
 }
 
+// GetFirmwareComponentsForNode gets all available firmware components for a node and return a list.
+func (p *ironicProvisioner) GetFirmwareComponentsForNode() ([]metal3api.FirmwareComponentStatus, error) {
+	ironicNode, err := p.getNode()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get node to retrieve firmware components")
+	}
+
+	// Get the components from Ironic via Gophercloud
+	var componentList []nodes.FirmwareComponent
+	var componentListErr error
+	componentList, componentListErr = nodes.ListFirmware(p.client, ironicNode.UUID).Extract()
+
+	if componentListErr != nil {
+		return nil, errors.Wrap(componentListErr,
+			fmt.Sprintf("could not get firmware components for node %s", ironicNode.UUID))
+	}
+	p.log.Info("retrieved firmware components for node", "node", ironicNode.UUID, "size", len(componentList))
+
+	// Setting to 2 since we only support bmc and bios
+	componentsInfo := make([]metal3api.FirmwareComponentStatus, 0, 2)
+
+	// Iterate over the list of components to extract their information and update the list.
+	for _, fwc := range componentList {
+		component := metal3api.FirmwareComponentStatus{
+			Component:          fwc.Component,
+			InitialVersion:     fwc.InitialVersion,
+			CurrentVersion:     fwc.CurrentVersion,
+			LastVersionFlashed: fwc.LastVersionFlashed,
+			UpdatedAt: metav1.Time{
+				Time: *fwc.UpdatedAt,
+			},
+		}
+		componentsInfo = append(componentsInfo, component)
+		p.log.Info("firmware component added for node", "component", fwc.Component, "node", ironicNode.UUID)
```

Comment:
This is going to be logged on every reconciler run, right? I'd rather log once when we discover the components (so move to updateHostFirmwareComponents)

Reply:
moved to updateHostFirmwareComponents